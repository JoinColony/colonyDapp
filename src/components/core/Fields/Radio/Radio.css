.delegate {
  position: absolute;
  opacity: 0;
}

.themeColorPicker {
  display: inline-block;
  margin: 8px 8px 0 0;
}

.themeColorPicker + .themeColorPicker {
  margin: 8px 8px 0 0;
}

.labelContainer {
  display: inline-block;
}

.radio {
  display: inline-block;
  height: 12px;
  width: 12px;
  position: relative;
  top: -3px; /* because the pseudo-radio is a bit smaller, let's shift it up a couple pixels */
  border: 2px solid var(--grey-4);
  border-radius: 50%;

  &::after {
    height: 3px;
    width: 3px;
    position: absolute;
    top: 50%;
    left: 50%;
    border-radius: 50%;
    background-color: var(--primary);
    content: '';
    transform: translate(-50%, -50%);
    transition: opacity 30ms ease-in;
    opacity: 0;
  }
}

.delegate:focus + .radio {
  box-shadow: 0 0 3px 1px var(--primary);
}

.main.stateIsChecked .radio {
  border-color: var(--primary);

  &::after {
    opacity: 1;
  }
}

.themeColorPicker.stateIsChecked .radio {
  border-color: var(--primary);
}

.themeColorPicker .radio {
  margin-right: 5px;
  height: 18px;
  width: 18px;
  border: 1px solid transparent;
  cursor: pointer;
}

.themeColorPicker .radio::after {
  height: 16px;
  width: 16px;
  border: 2px solid var(--grey-1);
  background-color: transparent;
}

.themeFakeCheckbox .radio {
  display: inline-block;
  flex: 0 0 auto;
  margin-right: 5px;
  height: 18px;
  width: 18px;
  position: relative;
  border: 1px solid var(--primary);
  border-radius: 2px;
  background-color: var(--temp-grey-2);
}

.themeFakeCheckbox .radio::after {
  opacity: 0;
}

/*
 * We copied over the checkmark styles here, as we can't compose in children
 * This is something we might want to solve using some sort of mixins eventually
 */
.themeFakeCheckbox .checkmark {
  display: inline-block;
  height: 80%;
  width: 80%;
  position: absolute;
  top: -20%;
  transform: rotate(45deg);
}

.themeFakeCheckbox.stateIsChecked .checkmark::before {
  height: 80%;
  width: 2px;
  position: absolute;
  right: 0;
  bottom: 0;
  background-color: var(--primary);
  content: '';
}

.themeFakeCheckbox.stateIsChecked .checkmark::after {
  height: 2px;
  width: 40%;
  position: absolute;
  right: 0;
  bottom: 0;
  background-color: var(--primary);
  content: '';
}

.themeButtonGroup {
  display: inline-block;
  padding: 6px 15px 0px 15px;
  border: 1px solid var(--grey-blue-1);
  background-color: var(--colony-white);
  white-space: nowrap;
  cursor: pointer;
}

.themeButtonGroup:first-of-type {
  border-top-left-radius: var(--radius-small);
  border-bottom-left-radius: var(--radius-small);
}

.themeButtonGroup + .themeButtonGroup:last-of-type {
  border-top-right-radius: var(--radius-small);
  border-bottom-right-radius: var(--radius-small);
}

.themeButtonGroup.stateIsChecked {
  border-color: var(--primary);
  background-color: var(--primary);
  color: var(--colony-white);
}

.themeButtonGroup + .themeButtonGroup {
  border-left: none;
}

.themeButtonGroup .radio {
  display: none;
}

.themeButtonGroup .labelContainer label {
  margin: 0;
  cursor: pointer;
}

/* Direction */

.directionHorizontal {
  display: inline-block;
  margin-right: 10px;

  & .radio {
    margin-right: 5px;
  }
}

.directionVertical {
  display: block;
  padding: 5px 0;

  & .radio {
    margin-right: 10px;
  }
}

/* State */

.stateIsDisabled {
  color: var(--grey-5);
  cursor: not-allowed;

  & .labelContainer label {
    cursor: not-allowed;
  }
}
