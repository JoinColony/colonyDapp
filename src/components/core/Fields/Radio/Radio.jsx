/* @flow */
import React, { Fragment } from 'react';
import nanoid from 'nanoid';
import { compose, lifecycle } from 'recompose';

import type { Node } from 'react';
import type { MessageDescriptor, MessageValues } from 'react-intl';

import { getMainClasses } from '~utils/css';

import InputLabel from '../InputLabel';
import asField from '../asField';

import styles from './Radio.css';

export type Appearance = {
  direction?: 'horizontal' | 'vertical',
  theme?: 'buttonGroup' | 'fakeCheckbox' | 'colorPicker',
};

type Props = {|
  /** Appearance object */
  appearance?: Appearance,
  /** If the input is checked */
  checked: boolean,
  /** Children to render in place of the default label */
  children?: Node,
  /** Disable the input */
  disabled?: boolean,
  /** Display the element without label */
  elementOnly?: boolean,
  /** Help text (will appear next to label text) */
  help?: string | MessageDescriptor,
  /** Values for help text (react-intl interpolation) */
  helpValues?: MessageValues,
  /** Label text */
  label: string | MessageDescriptor,
  /** Values for label text (react-intl interpolation) */
  labelValues?: MessageValues,
  /** Input field name (form variable) */
  name: string,
  /** Style object for the visible radio */
  radioStyle?: { [string]: string },
  /** @ignore Will be injected by `lifecycle` from `enhance` */
  inputId: string,
  /** @ignore Will be injected by `asField` */
  $id: string,
  /** @ignore Will be injected by `asField` */
  $error?: string,
  /** @ignore Will be injected by `asField` */
  $value?: string,
  /** @ignore Will be injected by `asField` */
  $touched?: boolean,
  /** @ignore Will be injected by `asField` */
  isSubmitting?: boolean,
  /** @ignore Will be injected by `asField` */
  formatIntl: (
    text: string | MessageDescriptor,
    textValues?: MessageValues,
  ) => string,
  /** @ignore Will be injected by `asField` */
  setValue: (val: any) => void,
  /** @ignore Will be injected by `asField` */
  setError: (val: any) => void,
  /** @ignore Standard input field property */
  onChange: Function,
|};

const displayName = 'Radio';

const Radio = ({
  appearance,
  checked,
  children,
  disabled,
  elementOnly,
  $error,
  help,
  helpValues,
  $id,
  inputId,
  label,
  labelValues,
  formatIntl,
  name,
  $value,
  $touched,
  radioStyle,
  setError,
  setValue,
  isSubmitting,
  connect,
  ...props
}: Props) => (
  <label
    className={getMainClasses(appearance, styles, {
      customChildren: !!children,
      isChecked: checked,
      isDisabled: !!disabled,
    })}
    htmlFor={elementOnly ? inputId : null}
  >
    <Fragment>
      <input
        className={styles.delegate}
        value={$value}
        name={name}
        type="radio"
        id={inputId}
        disabled={disabled}
        aria-checked={checked}
        aria-disabled={disabled}
        aria-invalid={!!$error}
        {...props}
      />
      <span className={styles.radio} style={radioStyle}>
        {!!appearance && appearance.theme === 'fakeCheckbox' && (
          <span className={styles.checkmark} />
        )}
      </span>
      {!elementOnly && !!label ? (
        <span className={styles.labelContainer}>
          <InputLabel
            appearance={{ direction: 'horizontal' }}
            label={label}
            help={help}
            inputId={inputId}
          />
        </span>
      ) : (
        label || children
      )}
    </Fragment>
  </label>
);

Radio.displayName = displayName;

Radio.defaultProps = {
  appearance: {
    direction: 'vertical',
  },
  disabled: false,
  elementOnly: false,
};

const enhance = compose(
  asField(),
  lifecycle({
    componentDidMount() {
      const inputId = nanoid();
      this.setState({ inputId });
    },
  }),
);

export default enhance(Radio);
