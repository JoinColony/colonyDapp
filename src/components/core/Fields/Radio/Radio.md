### Radio Inputs

```jsx
const { Formik } = require('formik');
const radioChoices = ['foo', 'bar', 'baz'];
<Formik
  onSubmit={values => console.log(values)}
  render={
    ({ handleSubmit, values }) => (
      <form onSubmit={handleSubmit}>
        <Heading appearance={{ size: 'medium' }} text="Standard Radio" />
        {radioChoices.map(radioChoice => (
          <Radio
            checked={values.radioInput === radioChoice}
            key={radioChoice}
            label={radioChoice}
            name="radioInput"
            value={radioChoice}
          />
        ))}
        <div style={{marginTop: '20px'}} />
        <Heading appearance={{ size: 'medium' }} text="Disabled Radio" />
        <Radio
          checked={values.radioInputDisabled === 'radioInputDisabled'}
          disabled
          label="Disabled radio"
          name="radioInputDisabled"
          value="disabledRadio"
        />
        <div style={{marginTop: '20px'}} />
        <Heading appearance={{ size: 'medium' }} text="Horizontal Radio" />
        {radioChoices.map(radioChoice => (
          <Radio
            appearance={{ direction: 'horizontal' }}
            checked={values.radioHorizontal === radioChoice}
            key={radioChoice}
            label={radioChoice}
            name="radioHorizontal"
            value={radioChoice}
          />
        ))}
        <div style={{marginTop: '20px'}} />
        <Heading appearance={{ size: 'medium' }} text="Fake Checkbox Radio" />
        {radioChoices.map(radioChoice => (
          <Radio
            appearance={{ theme: 'fakeCheckbox' }}
            checked={values.radioFakeCheckbox === radioChoice}
            key={radioChoice}
            label={radioChoice}
            name="radioFakeCheckbox"
            value={radioChoice}
          />
        ))}
        <div style={{marginTop: '20px'}} />
        <Heading appearance={{ size: 'medium' }} text="Button Group Radio" />
        {radioChoices.map(radioChoice => (
          <Radio
            appearance={{ theme: 'buttonGroup' }}
            checked={values.radioButtonGroup === radioChoice}
            key={radioChoice}
            label={radioChoice}
            name="radioButtonGroup"
            value={radioChoice}
          />
        ))}
        <div style={{marginTop: '20px'}} />
        <Heading appearance={{ size: 'medium' }} text="Color Picker Radio" />
        <Radio
          checked={values.radioColorPicker === 'colorPicker'}
          appearance={{ theme: 'colorPicker' }}
          name="radioColorPicker"
          label="And a color field"
          radioStyle={{ backgroundColor: 'blue' }}
          value="blue"
          help="Note - This doesn't provide an actual colorpicker input. This is still just a radio input"
        />
        <div style={{ marginTop: '20px' }} />
        <Button appearance={{ theme: 'primary' }} type="submit">And press me</Button>
      </form>
    )
  }
/>
```
