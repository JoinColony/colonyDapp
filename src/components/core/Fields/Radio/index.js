/* @flow */

export { default } from './Radio.jsx';

export type { Appearance } from './Radio.jsx';
