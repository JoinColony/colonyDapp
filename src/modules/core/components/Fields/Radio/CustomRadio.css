.input {
  height: 0;
  width: 0;
  opacity: 0;
}

.main {
  display: flex;
  align-items: center;
  padding: 7px;
  height: 41px;
  width: 100%;
  overflow: hidden;
  border: 1px solid transparent;
  border-radius: var(--radius-normal);
  background-color: color-mod(var(--temp-grey-blue-7) alpha(6%));
  color: var(--dark);
  cursor: pointer;
}

.themePrimary {
  composes: main;
}

.themeDanger {
  composes: main;
}

.themeGreyWithCircle {
  composes: main;
  border-color: color-mod(var(--action-secondary) alpha(25%));
  border-radius: var(--radius-tiny);
  background-color: color-mod(var(--temp-grey-blue-7) alpha(5%));
}

.themePrimary.stateIsChecked,
.themeGreyWithCircle.stateIsChecked {
  border: 1px solid var(--primary);
  background-color: color-mod(var(--primary) alpha(12%));
}

.themeDanger.stateIsChecked {
  border: 1px solid var(--pink);
  background-color: color-mod(var(--pink) alpha(12%));
}

.themePrimary:hover {
  background-color: color-mod(var(--primary) alpha(12%));
}

.themeDanger:hover {
  background-color: color-mod(var(--pink) alpha(12%));
}

.main.stateIsChecked .description {
  color: var(--temp-grey-blue-7);
  opacity: 0.85;
}

.main.stateIsDisabled .description {
  color: var(--temp-grey-blue-7);
  opacity: 0.85;
}

.main.stateIsDisabled .label {
  color: color-mod(var(--dark) alpha(65%));
}

.main.stateIsDisabled {
  border-color: transparent;
  background-color: var(--text-disabled);
  cursor: default;
}

.label {
  display: flex;
  align-items: center;
  font-size: var(--size-smallish);
  font-weight: var(--weight-bold);
  gap: 6px;
}

.description,
.description span {
  margin-left: auto;
  font-size: var(--size-tiny);
  font-weight: var(--weight-bold);
}

.icon {
  margin-right: 10px;
}

.content {
  display: flex;
  align-items: baseline;
  width: 100%;
  white-space: nowrap;
  gap: 6px;
}

.directionVertical .content {
  flex-direction: column;
  gap: 0;
}

.directionVertical .description {
  margin-left: 0;
}

.customRadioIcon {
  display: flex;
  flex-shrink: 0;
  justify-content: center;
  align-items: center;
  margin-right: 10px;
  height: 27px;
  width: 27px;
  border: 1px solid var(--temp-grey-blue-4);
  border-radius: 50%;
  background: var(--colony-white);
  box-shadow: inset 0px 2px 4px rgba(14, 37, 88, 0.07);
}

.checkedCustomRadioIcon {
  border-color: var(--primary);
}

.customRadioCheck {
  height: 7px;
  width: 7px;
  border-radius: 50%;
  background: var(--primary);
  box-shadow: var(--box-shadow-medium);
}
