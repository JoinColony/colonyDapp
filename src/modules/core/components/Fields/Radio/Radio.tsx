import React, { ReactNode, useRef } from 'react';
import nanoid from 'nanoid';

import { getMainClasses } from '~utils/css';

import InputLabel from '../InputLabel';
import asField from '../asField';
import { AsFieldEnhancedProps } from '../types';

import styles from './Radio.css';

export interface Appearance {
  direction?: 'horizontal' | 'vertical';
  theme?: 'buttonGroup' | 'fakeCheckbox' | 'colorPicker';
}

interface Props {
  /** Appearance object */
  appearance?: Appearance;
  /** If the input is checked */
  checked: boolean;
  /** Children to render in place of the default label */
  children?: ReactNode;
  /** Disable the input */
  disabled?: boolean;
  /** Style object for the visible radio */
  radioStyle?: { [k: string]: string };
  inputId?: string;
  value?: any;
}

const displayName = 'Radio';

const Radio = ({
  appearance = {
    direction: 'vertical',
  },
  checked,
  children,
  disabled = false,
  elementOnly = false,
  $error,
  help,
  inputId: inputIdProp,
  label,
  name,
  $value,
  radioStyle,
  /* eslint-disable @typescript-eslint/no-unused-vars */
  connect,
  $id,
  formatIntl,
  $touched,
  setError,
  setValue,
  isSubmitting,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ...props
}: Props & AsFieldEnhancedProps) => {
  const { current: inputId } = useRef<string>(inputIdProp || nanoid());
  return (
    <label
      className={getMainClasses(appearance, styles, {
        customChildren: !!children,
        isChecked: checked,
        isDisabled: !!disabled,
      })}
      htmlFor={elementOnly ? inputId : undefined}
    >
      <>
        <input
          className={styles.delegate}
          value={$value}
          name={name}
          type="radio"
          id={inputId}
          disabled={disabled}
          aria-checked={checked}
          aria-disabled={disabled}
          aria-invalid={!!$error}
          {...props}
        />
        <span className={styles.radio} style={radioStyle}>
          {!!appearance && appearance.theme === 'fakeCheckbox' && (
            <span className={styles.checkmark} />
          )}
        </span>
        {!elementOnly && !!label ? (
          <span className={styles.labelContainer}>
            <InputLabel
              appearance={{ direction: 'horizontal' }}
              label={label}
              help={help}
              inputId={inputId}
            />
          </span>
        ) : (
          label || children
        )}
      </>
    </label>
  );
};

Radio.displayName = displayName;

export default asField<Props>()(Radio);
