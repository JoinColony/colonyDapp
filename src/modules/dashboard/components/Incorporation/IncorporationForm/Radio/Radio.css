.radioWrapper > label {
  display: flex;
  align-items: center;
  margin-bottom: 9px;
  padding: 0 7px;
  height: 40px;
  width: 100%;
  position: relative;
  border: 1px solid color-mod(var(--action-secondary) alpha(25%));
  border-radius: 2px;
  background-color: color-mod(var(--action-secondary) alpha(5%));
  font-size: var(--size-smallish);
  font-weight: var(--weight-bold);
  color: var(--dark);
}

.radioWrapper > label > span {
  height: 27px;
  width: 27px;
  top: 0;
  border: 1px solid color-mod(var(--action-secondary) alpha(25%));
  background-color: var(--colony-white);

  &::after {
    height: 7px;
    width: 7px;
  }
}

.selected > label {
  border: 1px solid var(--primary);
  background-color: color-mod(var(--primary) alpha(12%));
}
