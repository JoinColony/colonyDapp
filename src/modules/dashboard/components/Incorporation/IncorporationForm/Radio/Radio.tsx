import React from 'react';
import { useIntl } from 'react-intl';
import classNames from 'classnames';

import { Radio as BaseRadio } from '~core/Fields';
import { Props as RadioProps } from '~core/Fields/Radio/Radio';

import styles from './Radio.css';

const displayName = 'dashboard.Incorporation.IncorporationForm.Radio';

const Radio = ({ name, checked, label, value }: RadioProps) => {
  const { formatMessage } = useIntl();
  const labelText = typeof label === 'object' ? formatMessage(label) : label;

  return (
    <div
      className={classNames(styles.radioWrapper, {
        [styles.selected]: checked,
      })}
    >
      <BaseRadio name={name} value={value} checked={checked} elementOnly>
        {labelText}
      </BaseRadio>
    </div>
  );
};

Radio.displayName = displayName;

export default Radio;
