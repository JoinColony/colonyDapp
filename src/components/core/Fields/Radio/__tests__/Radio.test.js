/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from 'testutils';

import Radio from '../Radio.jsx';

describe('Radio component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl(
      <Radio
        name="radioInput"
        help="halp"
        label="awesome"
        value="radioInput"
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
