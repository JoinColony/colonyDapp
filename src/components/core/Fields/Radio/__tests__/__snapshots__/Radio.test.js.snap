// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio component Renders initial component 1`] = `
<Component
  help="halp"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  label="awesome"
  name="radioInput"
  value="radioInput"
/>
`;
